keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,1050 @@
+external-provisioner via command line parameters. There is no need to
+introduce heuristics or other, more complex ways of changing the
+behavior (like extending the `CSIDriver` API).
+
+#### Available capacity vs. maximum volume size
+
+The CSI spec up to and including the current version 1.2 just
+specifies that ["the available
+capacity"](https://github.com/container-storage-interface/spec/blob/314ac542302938640c59b6fb501c635f27015326/lib/go/csi/csi.pb.go#L2548-L2554)
+is to be returned by the driver. It is left open whether that means
+that a volume of that size can be created.
+
+This KEP introduces the more precise definition of a "maximum volume
+size". Unless told otherwise, the external-provisioner assumes that
+the `GetCapacityResponse` is the sum of all currently available
+storage and then sets `CSIStorageByClass.AvailableCapacity` instead of
+`CSIStorageByClass.MaximumVolumeSize`. A CSI driver deployment can use
+`--capacity-is-maximum-volume-size=true` to change that, in which
+case only `CSIStorageByClass.MaximumVolumeSize` will be set.
+
+An extension of the CSI spec would be needed to enable a CSI driver to
+report both values.
+
+#### Without central controller
+
+This mode of operation is expected to be used by CSI drivers that need
+to track capacity per node and only support ephemeral inline volumes
+and/or persistent volumes with delayed binding that get provisioned
+by an external-provisioner instance that runs on the node where
+the volume gets provisioned (see the proposal for
+csi-driver-host-path in
+https://github.com/kubernetes-csi/external-provisioner/pull/367).
+
+The CSI driver has to implement the CSI controller service and its
+`GetCapacity` call. Its deployment has to add the external-provisioner
+to the daemon set and enable the per-node capacity tracking with
+`--enable-capacity=local`.
+
+The resulting `CSIStoragePool` objects then use a node selector for
+one node.
+
+#### With central provisioning
+
+Normally, external-provisioner gets deployed together with a CSI
+controller service in a stateful set. That allows drivers to
+support persistent volumes without late binding.
+
+When deployed with `--enable-capacity=central`, external-provisioner
+determines storage pools based on the topology information provided by
+the driver on each node and gets capacity for the resulting pools from
+the driver's controller service that it is connected to.
+
+This relies on a feature of kubelet and one additional requirement for CSI
+driver deployments:
+- Every topology segment that a CSI node service reports in response
+ to GetNodeInfo (for example, `example.com/zone`: `Z1` +
+ `example.com/rack`: `R3`) is copied into node labels by kubelet.
+- All keys must have the same prefix (`example.com` in that example)
+  and external-provisioner is given that prefix with the
+  `--topology-prefix=example.com` parameter.
+
+With that parameter, external-provisioner can reconstruct topology segments as follows:
```

Comment:
Can you give some examples of this?

What I'm not sure about, is how to generate the topology of a storage system that may span multiple segments, ie a storage pool that spans 2 nodes or 2 racks.

Reply:
Bummer :frowning_face:

While doing that I went back to the CSI spec and noticed that I had a slightly wrong interpretation of `NodeGetInfoResponse.Topology` in mind: I thought the driver reported where *storage* is accessible, not where the *node* is accessible.

In retrospect that makes sense of course, in particular when considering that the same driver might make storage available that has different constraints.

The effect is that the plan outlined here doesn't work. The alternative are either a CSI extension or additional parameters. This will require further thought and discussion, so for now let me propose that we take it out of the KEP, focus on local storage in the initial implementation, and then consider how to extend this for a revised alpha in 1.19:  0b200ea201bd70d6e120a9f81b8ef92a9a208293.

That's still a step forward. Node-local storage is where the lack of storage capacity tracking is most painful, too, so it makes sense to start there.